contracts/optimistic-ethereum/OVM/OVM_BondManager.sol

```diff
@@ -0,0 +1,201 @@
+
+    // Per pre-state root, store the number of state provisions that were made
+    // and how many of these calls were made by each user. Payouts will then be
+    // claimed by users proportionally for that dispute.
+    struct Rewards {
+        bool canClaim;
+        // Total number of `storeWitnessProvider` calls made
+        uint256 total;
+        // The sum of all values inside this map MUST be equal to the
+        // value of `totalClaims`
+        mapping(address => uint256) numClaims;
+    }
+    /// For each pre-state root, there's an array of witnessProviders that must be rewarded
+    /// for posting witnesses
+    mapping(bytes32 => Rewards) public witnessProviders;
+
+    /// Initializes with a ERC20 token to be used for the fidelity bonds
+    /// and with the Address Manager
+    constructor(ERC20 _token, address _libAddressManager)
+        Lib_AddressResolver(_libAddressManager)
+    {
+        owner = msg.sender;
+        token = _token;
+        // ovmFraudVerifier = resolve("OVM_FraudVerifier"); // TODO: Re-enable this
+        ovmCanonicalStateCommitmentChain = resolve("OVM_CanonicalStateCommitmentChain");
+    }
+
+    /// Adds `who` to the list of witnessProviders for the provided `preStateRoot`.
+    function storeWitnessProvider(bytes32 _preStateRoot, address who) public {
+        // The sender must be the transitioner that corresponds to the claimed pre-state root
+        address transitioner = address(iOVM_FraudVerifier(resolve("OVM_FraudVerifier")).getStateTransitioner(_preStateRoot));
+        require(transitioner == msg.sender, Errors.ONLY_TRANSITIONER);
+
+        witnessProviders[_preStateRoot].total += 1;
+        witnessProviders[_preStateRoot].numClaims[who] += 1;
+    }
+
+    /// Slashes + distributes rewards or frees up the sequencer's bond, only called by
+    /// `FraudVerifier.finalizeFraudVerification`
+    function finalize(bytes32 _preStateRoot, uint256 batchIndex, address publisher, uint256 timestamp) public {
+        require(msg.sender == resolve("OVM_FraudVerifier"), Errors.ONLY_FRAUD_VERIFIER);
+        require(witnessProviders[_preStateRoot].canClaim == false, Errors.ALREADY_FINALIZED);
+
+        // allow users to claim from that state root's
+        // pool of collateral (effectively slashing the sequencer)
+        witnessProviders[_preStateRoot].canClaim = true;
+
+        Bond storage bond = bonds[publisher];
+
+        // always slash 1 round of collateral, if possible
+        if (bond.locked >= requiredCollateral) {
+            bond.locked -= requiredCollateral;
+        }
+
+        // if the publisher has a pending withdrawal that's within the published
+        // block's challenge period
+        if (
+            bond.withdrawalTimestamp != 0 && 
+            bond.withdrawalTimestamp <= timestamp + forceInclusionPeriodSeconds &&
+            bond.withdrawing >= requiredCollateral
+        ) {
+            bond.withdrawing -= requiredCollateral;
```

Comment:
I think a simpler method here would be to set `bond.locked = 0` (no need to reduce exactly by `requiredCollateral`) and then assert that `bond.withdrawing <= bond.locked` within `finalizeWithdrawal` and that `bond.withdrawing - bond.locked >= requiredCollateral` for within `stake`

Reply:
As discussed, I'd like to avoid this optimization: Doing the changes as suggested seems to break my current tests.

Reply:
Ok we're just going to go with your current solution, marking this as resolved.